New: Adding Support for "// eslint-disable-line rule" style comments

chasingtheflow opened the pull request:
As discussed on eslint@googlegroups.com, this pull requests enables support for a trailing comment to "mute" or "ignore" a single line. For example, previously accomplishing this would have required something like:

```javascript
/*eslint-disable no-eval*/
var value = eval(value);
/*eslint-enable*/
``` 
Whereas now with `eslint-ignore` you can accomplish this more easily with a simple trailing comment:

```javascript
var value = eval(value);  /*eslint-ignore no-eval*/
```

Making it simpler and easier to mute known rule violations when necessary.

nzakas commented:
Thanks for the submission. Please take a moment to read over our contribution guidelines: http://eslint.org/docs/developer-guide/contributing.html

We'd need you to update your commit message, add tests, and sign our CLA.

lo1tuma commented on lib/eslint.js:
```diff
@@ -264,6 +264,11 @@ function modifyConfigsFromComments(ast, config, reportingConfig) {
 
+                    case "eslint-ignore":
+                        disableReporting(reportingConfig, { "line": comment.loc.start.line, "column": 0 }, Object.keys(parseListConfig(value)));
+                        enableReporting(reportingConfig, comment.loc.end, Object.keys(parseListConfig(value)));
```
This would allow you to ignore multiple lines.
Example: 
```js
var a = eval('alert()'), /* eslint-ignore no-eval




*/ b = eval('alert()');
```

lo1tuma commented:
You should also add tests and documentation for this feature.

chasingtheflow commented:
Hi Mathias,

I am working on updating my pull request to meet the required format and include tests, but it may be a few days before I have the opportunity to finish that up.

As for your observation about ignoring multiple lines, this isn’t really true in practice. For example, if we tested the following: 

function test (value) {
    "use strict";
    var val = eval(value); /*eslint-ignore no-eval
    */ var val2 = eval(value);

    return val + ' ' + val2;
}

Only the first eval would be ignored. The second eval would still cause an error. This is because the system parsing comments and setting “enable” and “disable” points actually utilizes the column value in addition to the line. So while the comment extends to a second line the actual “*/“ indicates where any “ignore” behavior ends. In theory you could have commented lines there, but they wouldn’t be tested by eslint as they’re inside a comment block. 

At least, this is how it is behaving for me. If you have an example which can demonstrate otherwise I’d love to see it and see whether there’s a work-around.

Best,

--
Billy Matthews
719.439.5484

lo1tuma commented:
Ah, ok that makes sense, thanks for your explanation.

nzakas commented:
Perhaps it would be better to require a single line comment for this?

chasingtheflow commented:
I’ve spent some time thinking about this, and can definitely see why you would bring it up. However, I would advocate against this for 2 reasons:

1. From what I can tell, all other eslint “comment directives” use the multi-line “/* */“ syntax, at least in the documentation, despite the fact that they are all designed to reside on a single line. Similarly, they would all also be subject to the same potential flaw which Mathias rightfully pointed out, which is that you could extend the comment to another line. Thought this wouldn’t impact the actual behavior, it could be cause for concern. However, seeing as all eslint comments seem to share this style, as well as its drawbacks, it seems that sticking with the convention would be worthwhile.

2. The complexity of breaking with the convention set forth by other comments would lead to unnecessary code complexity in implementing the feature. Maybe this is just schlep blindness on my part, as the potential implementer, as by leveraging the existing comment parsing system implementation is more straightforward. However, if we would need to filter /*eslint-ignore*/ style comments to only single line (e.g. “//“) style comments then this would add implementation complexity.

Given that the existing comment structure seems to favor the “/* */“ syntax, I think it would be both more confusing for end users and more difficult to implement this particular “comment directive” using only the single line (‘//‘) syntax.

But I’ll defer to the group for feedback and proceed accordingly.

Best,

--
Billy Matthews
719.439.5484

nzakas commented:
Hmm, I must say I disagree with both points. The convention is what it is because we have never had anything that should apply to just one line by default. So this would be a new convention.

It's also no more complex than verifying the type of comment before applying the same logic.

What I'm interested in is whether or not a single line comment makes more sense in this situation. It makes sense if it must always be a trailing comment on a line because it eliminates the confusion of having a multi line comment. It also make sense if we want a comment to appear before the line it should apply to.

So let's think long and hard about what the right thing to do here is, because we can't change it once released.

chasingtheflow commented:
Yes, I see your point. I really like that single line comments would enforce the comment being at the end of the line. To me eslint-ignore seems similar to eslint-disable, eslint-enable, which I would always envision as residing on a single line. Though I suppose what makes eslint-ignore different is that it is both expected to be on 1 line and expected to be at the end of the line. So enforcing single-line comments would certainly accomplish that. I also definitely agree that whatever we decide we cannot change it post release. I will update the pull request with Tests and modify it to enforce single-line comments only, unless anyone sees a benefit to allowing the block style?

nzakas commented:
Sounds like a good plan. 

I'm also wondering if maybe the comment name should be `eslint-disable-line`, so it's obvious that it's another way to disable a rule (or multiple rules) on a single line?

chasingtheflow commented:
Okay, so I've updated to use eslint-disable-line, added tests, and squashed the commits. But my editor has re-indented the whole file according to its own liking, and I'm not sure what's the best way to undo that ... but now the diff looks like a complete mess. Any advice on how I can clean that up? Thanks,

nzakas commented:
Yikes! You can try using `git reflog` to get to the state before you squashed.  See http://stackoverflow.com/questions/7457942/how-to-undo-a-git-merge-squash

chasingtheflow commented:
Hi Again, apologies for my newb git skills –– not something I use day-to-day. My latest commit has this cleaned up to the best of my abilities, but now I have 7 commits in this PR, where I know you only want 1. The last time I tried to "squash" these they got severely messed up when they merged together ... could someone give me advice on how to just include my latest commit? The diff now looks good to me however...

nzakas commented:
Looks like your latest commit still reformats a lot of code, so we probably don't want that.

At this point, I'd say your best bet is to create a new branch and manually copy over the parts you added.

chasingtheflow commented:
Okay, how does that look now?

nzakas commented:
Looks good! We just need two more things:

1. Can you modify the commit message to include "New:"
1. Can you update the configuration documentation to include this?

chasingtheflow commented:
Commit message modified, documentation added. :+1:

nzakas commented:
Great! So the last thing is to have you sign our CLA: http://eslint.org/CLA
This just gives us permission to use your code.

For future contributions please be sure to file an issue before sending a PR.

Thanks!

chasingtheflow commented:
Signed. Thanks!
